src/storage/lock_manager/deadlock.rs

```diff
@@ -23,83 +23,109 @@ use std::fmt::{self, Display, Formatter};
+    /// When checking the deadlock, if the ttl has elpased, the corresponding edge will be removed.
+    /// `last_detect_time` is the start time of the edge. `Detect` requests will refresh it.
+    // txn_ts => (lock_ts => (Vec<lock_hash>, last_detect_time))
+    wait_for_map: HashMap<u64, HashMap<u64, (Vec<u64>, time::Instant)>>,
+
+    now: time::Instant,
+
+    ttl: time::Duration,
 }
 
 impl DetectTable {
-    /// Returns the deadlock key hash which causes deadlock.
+    /// Creates a auto-expiring detect table.
+    pub fn new(ttl: time::Duration) -> Self {
+        Self {
+            wait_for_map: HashMap::default(),
+            now: time::Instant::now_coarse(),
+            ttl,
+        }
+    }
+
+    /// Returns the key hash which causes deadlock.
     pub fn detect(&mut self, txn_ts: u64, lock_ts: u64, lock_hash: u64) -> Option<u64> {
         let _timer = DETECTOR_HISTOGRAM_VEC.detect.start_coarse_timer();
         TASK_COUNTER_VEC.detect.inc();
 
-        // If txn_ts is waiting for lock_ts, it won't cause deadlock.
+        self.now = time::Instant::now_coarse();
+        // If `txn_ts` is waiting for `lock_ts`, it won't cause deadlock.
         if self.register_if_existed(txn_ts, lock_ts, lock_hash) {
             return None;
         }
 
-        // Memorize the searching path.
-        let mut visited: HashSet<u64> = HashSet::default();
-        if let Some(deadlock_key_hash) = self.do_detect(&mut visited, txn_ts, lock_ts) {
+        if let Some(deadlock_key_hash) = self.do_detect(txn_ts, lock_ts) {
             return Some(deadlock_key_hash);
         }
         self.register(txn_ts, lock_ts, lock_hash);
         None
     }
 
-    /// Checks if there is a path from `wait_for_ts` to `txn_ts`.
-    fn do_detect(&self, visited: &mut HashSet<u64>, txn_ts: u64, wait_for_ts: u64) -> Option<u64> {
-        visited.insert(wait_for_ts);
-        if let Some(wait_for) = self.wait_for_map.get(&wait_for_ts) {
-            for (lock_ts, lock_hashes) in wait_for {
-                if *lock_ts == txn_ts {
-                    return Some(lock_hashes[0]);
-                }
-                if visited.contains(lock_ts) {
-                    continue;
-                }
-                if let Some(deadlock_key_hash) = self.do_detect(visited, txn_ts, *lock_ts) {
-                    return Some(deadlock_key_hash);
+    /// Checks if there is a edge from `wait_for_ts` to `txn_ts`.
+    fn do_detect(&mut self, txn_ts: u64, wait_for_ts: u64) -> Option<u64> {
+        let now = self.now;
+        let ttl = self.ttl;
+
+        // Memorize the searched vertexes.
+        let mut visited: HashSet<u64> = HashSet::default();
+        let mut stack = vec![wait_for_ts];
+        while let Some(wait_for_ts) = stack.pop() {
+            if let Some(wait_for) = self.wait_for_map.get_mut(&wait_for_ts) {
+                // Remove expired edges.
+                wait_for.retain(|_, (_, time)| now.duration_since(*time) < ttl);
```

Comment:
When the transaction is committed or rollbacked, it will send `CleanUp` request to it. If the waiter is waked up, it will send `CleanUpWaitFor` request to it. If these requests fail to send, it's possible to leave some entries in the table.

The purpose of the expiration mechanism is to prevent the result of  `Detect`  affected by left entries. It can't clean up all expired entries, but I think it doesn't matter.

Reply:
Got it. By the way, in the future do we need to prevent these stale entries cost too much memory?

Reply:
Better to have, but I can't think of a good way to solve it except a timer.(I don't like timer)

But as long as the leader of the region containing `b""` is changed, all detect table will be clear. So if it is a problem, we can change the leader manually.

Reply:
Or scan the detect table if the size is too big?

Reply:
Adds it to `detect`? The `ACTIVE_EXPIRE_THRESHOLD` is 10w?
```rust
    fn active_expire(&mut self) {
        if self.wait_for_map.len() < ACTIVE_EXPIRE_THRESHOLD {
            return;
        }

        let now = self.now;
        let ttl = self.ttl;
        for (_, wait_for) in self.wait_for_map.iter_mut() {
            wait_for.retain(|_, (_, time)| now.duration_since(*time) < ttl);
        }
        self.wait_for_map.retain(|_, wait_for| wait_for.len() > 0);
    }
```

Reply:
If the cleanup operation is too heavy, maybe a timer is needed to avoid causing too much latency when the requests do the lazy cleanup.

Reply:
Is it possible that wait_for_map is still larger than 10W after active expire?

Reply:
Yes, if there are so many active entries. I think it is undeserved to do it, because it's almost impossible to leave many entries.

Reply:
Sounds like a good idea. Function as a timer but it is not a timer.

Reply:
@coocood @MyonKeminta PTAL again.